### PR TITLE
feat: I/O safety for 'sys/inotify'

### PR DESCRIPTION
**Pull request by SteveLauC**
#### What this PR does:

1. Changes the `fd` field of `struct Inotify` from `RawFd` to `OwnedFd`
2. Changes the interfaces of functions in the `impl Inotify {}`
   
    > The type of `self` changes from `Self` to `&mut Self`. 
    
    From:

   ```rust
   pub fn add_watch<P: ?Sized + NixPath>(
         self,
         path: &P,
         mask: AddWatchFlags,
   ) -> Result<WatchDescriptor> 

   pub fn rm_watch(self, wd: WatchDescriptor) -> Result<()>

   pub fn read_events(self) -> Result<Vec<InotifyEvent>>
   ```

   To:

   ```rust
   pub fn add_watch<P: ?Sized + NixPath>(
        &mut self,
        path: &P,
        mask: AddWatchFlags,
    ) -> Result<WatchDescriptor>

   pub fn rm_watch(&mut self, wd: WatchDescriptor) -> Result<()>

   pub fn read_events(&mut self) -> Result<Vec<InotifyEvent>>
   ```
  

   In the previous implementation, these functions can take `self` by value as `struct Inotify` [was `Copy`](https://docs.rs/nix/latest/nix/sys/inotify/struct.Inotify.html#impl-Copy-for-Inotify). With the changes in `1` applied, `struct Inotify` is no longer `Copy`, so we have to take `self` by reference.

-------

Blocks until the merge of #1863 as this PR needs `read(2)` to be I/O-safe.

### CODE REVIEW

File: src/sys/inotify.rs
Comment (SteveLauC) [inline src/sys/inotify.rs:243]: This trait impl extracts the `RawFd` from `struct Inofity`, IMO, this is used to `close(2)` the file descriptor representing the `inotify` instance, which is no longer needed as `OwnedFd` will drop it automatically.

**Comment by asomers** [comment]
The CI failure looks directly related to this PR.

**Comment by SteveLauC** [comment]
> The CI failure looks directly related to this PR.

Yes, that's why this PR needs to be blocked until the merge of #1863

**Comment by SteveLauC** [comment]
@asomers I have fixed the error (by using the old `RawFd` `read(2)` interface) and this is ready for review.

File: src/sys/inotify.rs
Comment (asomers) [inline src/sys/inotify.rs:155]: Why `&mut self` and not `&self`?

Comment (SteveLauC) [reply src/sys/inotify.rs:155]: Semantically, inotify_add_watch modifies the existing inotify instance, though this can not be reflected in the implementation of this binding.

[Man 2 inotify_add_watch](https://man7.org/linux/man-pages/man2/inotify_add_watch.2.html)

Comment (asomers) [reply src/sys/inotify.rs:155]: I consider this is a case of "internal mutability". The correctness of the the method doesn't depend on it having exclusive access to the the structure, so we can remove the `mut`.

Comment (SteveLauC) [reply src/sys/inotify.rs:155]: Get it, this can also be applied to rm_watch and read_events

**Review by asomers — APPROVED** [review]
bors r+

**Comment by VorpalBlade** [comment]
It seems this broke monitoring Inotify with Epoll since it removed the AsRawFd but didn't implement any other way to combine it will Epoll (or plain Poll for that matter). This makes it impossible to monitor both inotify files and some other FDs at the same time.

See also issue #1998

**Pull request merged**
